### PR TITLE
feature: Update Google Sign-In to Credential Manager

### LINKED ISSUE #107: SignIn: Switch from Google Sign-In to Credential Manager

### PR DESCRIPTION
**Pull request by aaisel**
# Description

This update refactors the Google Sign-In flow to use the new Credential Manager API. This change replaces the deprecated Google Sign-In methods and aligns with best practices for user authentication. Additionally, it introduces a logout button to enhance user control over their session.

Fixes #107 

# How Has This Been Tested?

### Android Tests

The tests found in the  `SignInScreenTest` file have been updated to match the current implementation of the `SignInScreen` Composable. Furthermore, an additional test to verify the functionality of the logout button was added in `ProfileScreenTest` .
  
# Dependencies

Necessary dependencies are added for `androidx-credentials`, `androidx-credentials-play-services-auth`, and `google-id`.

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged in downstream modules, no conflicts with main

### CODE REVIEW

File: app/src/main/java/com/github/onlynotesswent/ui/user/PublicProfile.kt [hunk 1]
```diff
@@ -201,6 +206,12 @@ fun TopProfileBar(
             Icon(imageVector = Icons.AutoMirrored.Outlined.ArrowBack, contentDescription = "Back")
           }
         }
+      },
+      actions = {
```
Comment (Roshan-y) [inline app/src/main/java/com/github/onlynotesswent/ui/user/PublicProfile.kt:210]: right now the log out button shows for all profile screen (including when you are looking at other users). To fix this you can just if the val "includeBackButton" and only display the log out if this val is false. You could also update the name of this val to "isUserProfileScreen".

**Review by Roshan-y — CHANGES_REQUESTED** [review]
Overall very well documented and modularised. The code follow's good practices. I have a few suggestions for GoogleCredSignIn which will make for cleaner code. Also logout button is displayed in unnecessary screens which I also commented about.

File: app/src/main/java/com/github/onlynotesswent/model/authentication/GoogleCredSignIn.kt [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+ *
+ * @param ctx The context to use for the sign-in request
+ * @param credentialManager The CredentialManager instance to use
+ * @param serverClientId The server client ID to use for the sign-in request
+ * @param launcher The launcher to use for the fallback Google sign-in method
+ */
+class GoogleCredSignIn(
+    private val ctx: Context,
+    private val credentialManager: CredentialManager,
+    private val serverClientId: String,
```
[the remark this answers is not in the document]
Comment (aaisel) [reply app/src/main/java/com/github/onlynotesswent/model/authentication/GoogleCredSignIn.kt:37]: I had some issues with testing when doing that. Will modify if the problem gets fixed

File: app/src/main/java/com/github/onlynotesswent/model/authentication/GoogleCredSignIn.kt [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+    // Retrieve user's available credentials or prompt them to sign in
+    CoroutineScope(Dispatchers.IO).launch {
+      try {
+        // Attempt to get the Google ID credential
+        val result = credentialManager.getCredential(context = ctx, request = request)
+        handleSignIn(callback, result)
+      } catch (e: Exception) {
+        Log.e(TAG, "Error getting credential", e)
+        // Exception can happen when the device has never had a Google account before
+        // In this case, reattempt the sign-in using traditional Google sign-in
```
[the remark this answers is not in the document]
Comment (aaisel) [reply app/src/main/java/com/github/onlynotesswent/model/authentication/GoogleCredSignIn.kt:68]: Thanks

File: app/src/main/java/com/github/onlynotesswent/model/authentication/GoogleCredSignIn.kt [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+import com.google.android.libraries.identity.googleid.GoogleIdTokenCredential
+import com.google.android.libraries.identity.googleid.GoogleIdTokenParsingException
+import java.security.MessageDigest
+import java.util.UUID
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.withContext
+
+private const val TAG = "GoogleCredSignIn"
```
[the remark this answers is not in the document]
Comment (aaisel) [reply app/src/main/java/com/github/onlynotesswent/model/authentication/GoogleCredSignIn.kt:24]: Nice

File: app/src/main/java/com/github/onlynotesswent/ui/user/PublicProfile.kt (hunk 1 above)
Comment (aaisel) [reply app/src/main/java/com/github/onlynotesswent/ui/user/PublicProfile.kt:210]: Nice catch

File: app/src/main/java/com/github/onlynotesswent/model/authentication/GoogleCredSignIn.kt [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+    }
+  }
+}
+
+/**
+ * Generate a nonce for the Google sign-in request
+ *
+ * @return A hashed nonce
+ */
+fun generateNonce(): String {
```
[the remark this answers is not in the document]
Comment (aaisel) [reply app/src/main/java/com/github/onlynotesswent/model/authentication/GoogleCredSignIn.kt:117]: I will only concatenate a few, to maintain it readable and also optimize it

File: app/src/main/java/com/github/onlynotesswent/model/authentication/GoogleCredSignIn.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (Roshan-y) [reply app/src/main/java/com/github/onlynotesswent/model/authentication/GoogleCredSignIn.kt:36]: Yah it might be better to keep it as it is for testing.

**Review by Roshan-y — APPROVED** [review]

**Pull request merged**